client/consensus/common/src/level_monitor.rs

```diff
@@ -0,0 +1,325 @@
+	/// then the limit is enforced by chosing one (or more) blocks to remove.
+	///
+	/// The removal strategy is driven by the block freshness.
+	///
+	/// A block freshness is determined by the most recent leaf freshness descending from the block
+	/// itself. In other words its freshness is equal to its more "fresh" descendant.
+	///
+	/// The least "fresh" blocks are eventually removed.
+	pub fn enforce_limit(&mut self, number: NumberFor<Block>) {
+		if self.import_counter == Zero::zero() {
+			// First call detected, restore using the backend persistent information.
+			self.restore();
+		}
+
+		let level_len = self.levels.get(&number).map(|l| l.len()).unwrap_or_default();
+		if level_len < self.level_limit {
+			return
+		}
+
+		log::debug!(
+			target: "parachain",
+			"Detected leaves overflow at height {}, removing obsolete blocks",
+			number,
+		);
+
+		let mut leaves = self.backend.blockchain().leaves().unwrap_or_default();
+		// Sort leaves by freshness (less fresh first).
+		leaves.sort_unstable_by(|a, b| self.freshness.get(a).cmp(&self.freshness.get(b)));
+
+		// This may not be the most efficient way to remove **multiple** entries, but is the easy
+		// one :-). Should be considered that in "normal" conditions the number of blocks to remove
+		// is 0 or 1, it is not worth to complicate the code too much. One condition that may
+		// trigger multiple removals (2+) is if we restart the node using an existing db and a
+		// smaller limit wrt the one previously used.
+		let remove_count = level_len - self.level_limit + 1;
+		(0..remove_count).all(|_| {
+			self.find_target(number, &leaves).map_or(false, |target| {
+				self.remove_target(target, &leaves);
+				true
+			})
+		});
+	}
+
+	// Helper function to find the best candidate to be removed.
+	//
+	// Given a set of blocks with height equal to `number` (potential candidates)
+	// 1. For each candidate fetch all the leaves that are descending from it.
+	// 2. Set the candidate freshness equal to the fresher of its descending leaves.
+	// 3. The target is set as the candidate that is less fresh.
+	//
+	// Input `leaves` are assumed to be already ordered by "freshness" (less fresh first).
+	//
+	// Returns the index of the target fresher leaf within `leaves` and the route from target to
+	// such leaf.
+	fn find_target(
+		&self,
+		number: NumberFor<Block>,
+		leaves: &[Block::Hash],
+	) -> Option<TargetInfo<Block>> {
+		let mut target_info: Option<TargetInfo<Block>> = None;
+		let blockchain = self.backend.blockchain();
+		let best_hash = blockchain.info().best_hash;
+
+		let level = match self.levels.get(&number) {
+			Some(level) => level,
+			None => return None,
+		};
+
+		for blk_hash in level.iter().filter(|hash| **hash != best_hash) {
+			// Search for the fresher leaf information for this block
+			let candidate_info =
+				leaves.iter().enumerate().rev().find_map(|(leaf_idx, leaf_hash)| {
```

Comment:
We should filter out leaves that we already successfully found with `route.retracted().is_empty()`.

And why not start with the oldest leaves? Wouldn't that make more sense?

Reply:
> We should filter out leaves that we already successfully found with route.retracted().is_empty().

Good catch

> And why not start with the oldest leaves? Wouldn't that make more sense?

We start from the fresher leaves to stop as soon as we find one descending from the node under processing (we are looking for its fresher descendant).

Reply:
> We start from the fresher leaves to stop as soon as we find one descending from the node under processing (we are looking for its fresher descendant).

Yeah I got this, my question was more, wouldn't it be faster to directly start with the oldest (less freshest leave)? I mean yes, it always depends on how the blocks on the level are ordered in relation to the leaves. However, in theory we should be able to find the oldest leave directly for the first block, which on the other way around, as you are currently doing it, wouldn't be possible.